common/src/main/java/com/microsoft/identity/common/exception/CommonCoreBaseException.java

```diff
@@ -0,0 +1,149 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.exception;
+
+import com.microsoft.identity.common.adal.internal.net.HttpWebResponse;
+import com.microsoft.identity.common.adal.internal.util.StringExtensions;
+import com.microsoft.identity.common.adal.internal.util.HashMapExtensions;
+
+import org.json.JSONException;
+
+import java.util.HashMap;
+import java.util.List;
+
+public class CommonCoreBaseException extends Exception {
+    private String mErrorCode;
+
+    /**
+     * Default constructor.
+     */
+    CommonCoreBaseException() {
+    }
+
+    /**
+     * Initiates the detailed error code.
+     *
+     * @param errorCode The error code contained in the exception.
+     */
+    public CommonCoreBaseException(final String errorCode) {
+        mErrorCode = errorCode;
+    }
+
+    /**
+     * Initiates the {@link CommonCoreBaseException} with error code and error message.
+     *
+     * @param errorCode    The error code contained in the exception.
+     * @param errorMessage The error message contained in the exception.
+     */
+    public CommonCoreBaseException(final String errorCode, final String errorMessage) {
+        super(errorMessage);
+        mErrorCode = errorCode;
+    }
+
+    /**
+     * Initiates the {@link CommonCoreBaseException} with error code, error message and throwable.
+     *
+     * @param errorCode    The error code contained in the exception.
+     * @param errorMessage The error message contained in the exception.
+     * @param throwable    The {@link Throwable} contains the cause for the exception.
+     */
+    public CommonCoreBaseException(final String errorCode, final String errorMessage,
+                  final Throwable throwable) {
+        super(errorMessage, throwable);
+        mErrorCode = errorCode;
+    }
+
+    /**
+     * @return The error code for the exception, could be null. {@link CommonCoreBaseException} is the top level base exception, for the
+     * constants value of all the error code.
+     */
+    public String getErrorCode() {
+        return mErrorCode;
+    }
+
+    /**
+     * {@inheritDoc}
+     * Return the detailed description explaining why the exception is returned back.
+     */
+    @Override
+    public String getMessage() {
+        if (!StringExtensions.isNullOrBlank(super.getMessage())) {
+            return super.getMessage();
+        }
+
+        return "";
```

Comment:
I'm OK with returning `null` here, unless there's a reason `""` is preferable to you

Reply:
By .NET conventions, APIs should not return null when the return value type is string. I do not know what Java says about this.

Reply:
Java doesn't share this convention -- the default implementation of `Throwable#getMessage()` returns `null` ([docs](https://developer.android.com/reference/java/lang/Throwable.html#getMessage()))

Reply:
fixed